astropy/modeling/core.py

```diff
@@ -1224,8 +1392,114 @@ def _param_sets(self, raw=False):
             psets[:] = values
             return psets
 
+        # TODO: Returning an array from this method may be entirely pointless
+        # for internal use--perhaps only the external param_sets method should
+        # return an array (and just for backwards compat--I would prefer to
+        # maybe deprecate that method)
+
         return np.array(values)
 
+    def _prepare_input_units(self, inputs):
+        using_quantities = self._using_quantities
+        using_quantities |= any(isinstance(input_, Quantity)
+                                for input_ in inputs)
+
+        if using_quantities:
+            inputs = self._prepare_variable_units(inputs)
+
+        return inputs, using_quantities
+
+    def _prepare_output_units(self, inputs, outputs):
+        if not self._using_quantities:
+            return outputs
+        else:
+            return self._prepare_variable_units(inputs, outputs)
+
+    def _prepare_variable_units(self, inputs, outputs=None):
+        if outputs is None:
+            unit_specs = self.input_units
+            var_names = self.inputs
+            n_vars = self.n_inputs
+            vars_ = inputs
+        else:
+            unit_specs = self.output_units
+            var_names = self.outputs
+            n_vars = self.n_outputs
+            vars_ = outputs
+
+        if unit_specs is None:
+            unit_specs = (None,) * n_vars
+        elif not isinstance(unit_specs, tuple):
+            unit_specs = (unit_specs,) * n_vars
+
+        def to_unit(value, unit_spec):
+            # Go from an element in the output_units attributes to an actual
+            # Unit object
+            if isinstance(unit_spec, UnitBase):
```

Comment:
Unfortunately, currently this would not cover logarithmic units... Currently, in `Unit`, the [following] ( https://github.com/astropy/astropy/blob/master/astropy/units/core.py#L1744) is used to circumvent this:
``` 
if hasattr(s, '_get_physical_type_id'):
```